event manager: Implement event manager

zycz opened the pull request:
Moves event manager with sample from sdk-nrf to zephyr. 

Event Manager is a piece of software that supports development of consistent, modular, event-based application. In an event-based application, parts of the application functionality are separated into isolated modules that communicate with each other using events. Events are submitted by modules and other modules can subscribe and react to them.

This PR aims to:
-Initiate the discussion.
-Push the event manager as-is but once comments are taken we will move forward. This will be done in another task.
-Check if all applications already using event manager work fine with event manager placed in zephyr
-Introduce Event manager code into Zephyr

zycz committed:
event manager: implementation of event manager

Add implementation and documentation of Event manager.

Signed-off-by: Jan Zyczkowski <jan.zyczkowski@nordicsemi.no>

zycz committed:
event manager: event manager sample

Add sample aplication of event manager with documentation to it.
Aplication shows how events are submited and received .

Signed-off-by: Jan Zyczkowski <jan.zyczkowski@nordicsemi.no>

carlescufi commented on include/event_manager.h:
```diff
@@ -0,0 +1,307 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
```
Please replace with `SPDX-License-Identifier: Apache-2.0`

carlescufi commented on include/profiler.h:
```diff
@@ -0,0 +1,297 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
```
Please replace with `SPDX-License-Identifier: Apache-2.0`

carlescufi commented on subsys/event_manager/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+#
+# Copyright (c) 2018 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
```
Please replace with `SPDX-License-Identifier: Apache-2.0`

carlescufi commented on subsys/event_manager/Kconfig:
```diff
@@ -0,0 +1,64 @@
+#
+# Copyright (c) 2018 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
```
Please replace with `SPDX-License-Identifier: Apache-2.0`

carlescufi commented on subsys/event_manager/event_manager.c:
```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
```
Please replace with `SPDX-License-Identifier: Apache-2.0`

carlescufi commented on subsys/event_manager/event_manager_priv.h:
```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
```
Please replace with `SPDX-License-Identifier: Apache-2.0`

carlescufi commented on subsys/event_manager/event_manager_shell.c:
```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
```
Please replace with `SPDX-License-Identifier: Apache-2.0`

nordic-krch commented:
Couple of comments:
- `include/profiler.h` didn't catch if it is event_manager specific or something more generic.
- there are references to NCS in the documentation
- maybe sample should be moved to `samples/application_development`?

As for event manager design, there are few things that could be added later on:
- option to use dedicated thread instead of system work queue.
- option to not use heap. I think that `mpsc_pbuf` (MPSC packet buffer) would suit well for that purpose.

zycz commented:
> * `include/profiler.h` didn't catch if it is event_manager specific or something more generic.
Profiler is a tool for monitoring events both from event manager and custom ones.  https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/libraries/others/profiler.html#profiler
We add profiler.h in this PR to ensure event manager will still be able to work with the profiler.

> * there are references to NCS in the documentation
> * maybe sample should be moved to `samples/application_development`?

Ok, I will fix it.

nordic-krch commented on subsys/event_manager/event_manager_priv.h:
```diff
@@ -0,0 +1,256 @@
+		},													\
+		.init_log_enable		= init_log_en,								\
+		.log_event			= log_fn,								\
+		.ev_info			= ev_info_struct,							\
```
it should be assigned only if profiler is enabled. Now event infos are created even when not used. That should also allow to clean up double defines for some macros, eg. `MEM_ADDRESS_LABEL`

nordic-krch commented on samples/application_development/event_manager/src/events/control_event.c:
```diff
@@ -0,0 +1,23 @@
+EVENT_TYPE_DEFINE(control_event,
+		  true,
+		  NULL,
+		  &control_event_info);
```
i don't like that event structure gets suffix and user must somehow know that after calling `EVENT_INFO_DEFINE(control_event,..)` it must use `control_event_info`. Imo, event info name should be explicit so that user does something like:
```
EVENT_INFO_DEFINE(control_event_info,
		  ENCODE(),
		  ENCODE(),
		  profile_control_event);

EVENT_TYPE_DEFINE(control_event,
		  true,
		  NULL,
		  &control_event_info);

nordic-krch reviewed (commented):
few more comments, what about test? Would be good to get that in as well.

zycz commented on boards/riscv/tlsr9518adk80d/doc/index.rst:
```diff
@@ -81,6 +81,7 @@ The following example projects are supported:
 - samples/hello_world
 - samples/synchronization
 - samples/philosophers
+- samples/event_manager
```
Are those pages for each board autogenerated? I didn't have to add samples/event_manager manualy. If yes why it did not change for sample/application_development/event_manager automatically?

nashif replied:
I am not sure why are we adding this for this board specifically?  (and why the board is listing all the samples it supports in the first place)

hongshui3000 commented:
Very useful, hope to be merged. :)

JordanYates reviewed (commented):
My thoughts on the general themes.

The inability to register new event subscribers at runtime is not great. #2746 has seen some renewed interest recently, and having dynamic modules unable to hook into events would be a big limitation.

There is no inbuilt integration with the existing Zephyr synchronization paradigm (poll). If I have a thread that just wants to run some operations on a new GPS sample that comes in, I not only need to raise the appropriate signals in the callback, but also manually copy out the event to some temporary storage, as the buffer doesn't exist after the subscribers have been notified. This is a larger downside if your event is a new buffer of 128 16bit 3-axis accelerometer samples (768 bytes).

Configuring `CONFIG_HEAP_MEM_POOL_SIZE` (via guess and check) for each application will never be fun. For a robust system, it will need to be at least as large as the sum of all the individual event sizes. This information is known at the locations where events are generated, but I can't see a way to make this any easier.

Less important, and not necessarily in scope for what the subsys is trying to achieve, but it is not possible to query the latest value of an event. For example, code may not necessarily care how often the battery voltage changes, or want to be notified each time a measurement is obtained, but simply query what the current state is.

nordic-krch commented:
> modules themselves can't explicitly state "I now care about these events"

It can be done locally in the listener by simply checking if given event is enabled for me. I guess it could be moved to internals of event manager, use some bit mask and CLZ to quickly iterate over enabled listeners. It is only the optimization. 

Regarding collecting sensor samples to a database which could be shared between the modules, i don't think it is the scope of event manager. Event manager just distributes event to loosely coupled modules. One of those modules can be a database that collects the samples.

Regarding use of heap, there are discussions to add an option to use dedicated buffer space instead of heap (e.g. mpsc_pbuf could be used which serves the event manager case - multiple producers, single consumer).

It would be good to see what alternative could be used.

carlescufi commented:
@JordanYates @nordic-krch it would be great if we could move this review forward.

carlescufi commented:
@zycz could you please rebase against the latest master and push?

nashif commented on doc/reference/kernel/index.rst:
```diff
@@ -126,3 +126,4 @@ These pages cover other kernel services.
    other/version.rst
    other/fatal.rst
    other/thread_local_storage.rst
+   other/event_manager.rst
```
Why are you adding this under kernel/other? This should be alongside other subsystems maybe? The other/ under kernel is historical and needs some cleanup, so lets not add more things here.

zycz replied:
Thank you for your review. Could you point me where should this documentation in zephyr be placed?

carlescufi replied:
Please place it in doc/reference/event_manager

nashif commented on doc/reference/kernel/other/event_manager.rst:
```diff
@@ -0,0 +1,396 @@
+To use Event Manager, you must enable the following Kconfig options:
+
+* :kconfig:`CONFIG_EVENT_MANAGER` - This option enables the Event Manager.
+* :kconfig:`CONFIG_LINKER_ORPHAN_SECTION_PLACE` - This option enables orphan memory sections used by the Event Manager.
```
why not fix the orphan sections instead? and, just looking at the code, this Kconfig option seems to be unused and a leftover.

zycz replied:
Sorry, did not understand your question. If you are asking whether we are using orphan sections, the answer is yes. If you disable this option you wouldn't be able to build sample.

nashif commented on include/event_manager/profiler.h:
```diff
@@ -0,0 +1,297 @@
+/** @brief Buffer required for data that is sent with the event.
+ */
+struct log_event_buf {
+#ifdef CONFIG_PROFILER
```
where is this CONFIG_PROFILER defined? This is a very generic kconfig to use and it is not defined in this PR, did you mean to rename it to something else?

zycz replied:
As agreed in previous comments all data related to profiler will be removed from event manager.

nashif commented on boards/riscv/tlsr9518adk80d/doc/index.rst:
```diff
@@ -81,7 +81,7 @@ The following example projects are supported:
 - samples/synchronization
 - samples/philosophers
-- samples/event_manager
+- samples/application_development/event_manager
```
probably we need to have another move, this is being introduced as a subsystem, so it needs to end up in samples/subsys/event_manager. Also, I am not sure why you have it in this file

zycz replied:
Ok, it can be moved there.

nashif commented on samples/application_development/event_manager/prj.conf:
```diff
@@ -0,0 +1,19 @@
+CONFIG_EVENT_MANAGER=y
+CONFIG_LINKER_ORPHAN_SECTION_PLACE=y
+CONFIG_HEAP_MEM_POOL_SIZE=2048
+CONFIG_REBOOT=y
```
why do we need this?

zycz replied:
We run sys_reboot if there is no enough memory to allocate event.

hongshui3000 replied:
Why not wait for the memory to be released? Is it due to support for single thread?

zycz replied:
We cannot drop events as it would lead to unexpected situations. If such problems happens it is safer to reboot.

nashif commented on samples/application_development/event_manager/prj.conf:
```diff
@@ -0,0 +1,19 @@
+#
+# Copyright (c) 2019 Nordic Semiconductor ASA
```
2021?

nashif reviewed (dismissed):
Can you please add a test for this new subsystem?

JordanYates commented:
> Regarding collecting sensor samples to a database which could be shared between the modules, i don't think it is the scope of event manager. Event manager just distributes event to loosely coupled modules. One of those modules can be a database that collects the samples.

While in a less constrained environment that would definitely make sense, many micro's and applications aren't going to have that much spare RAM to throw around.

My point is that to ensure sane operation of this subsystem, you are practically forced to increase the heap size by the sum of the sizes of all the individual events. Therefore, you are already adding enough memory to your application to implement this functionality for free (minus some synchronization primitives).

pdunaj commented:
> > Regarding collecting sensor samples to a database which could be shared between the modules, i don't think it is the scope of event manager. Event manager just distributes event to loosely coupled modules. One of those modules can be a database that collects the samples.
> 
> While in a less constrained environment that would definitely make sense, many micro's and applications aren't going to have that much spare RAM to throw around.
> 
> My point is that to ensure sane operation of this subsystem, you are practically forced to increase the heap size by the sum of the sizes of all the individual events. Therefore, you are already adding enough memory to your application to implement this functionality for free (minus some synchronization primitives).

Hi @JordanYates , thank you for all the comments. EM was designed for a specific application. We use it in other applications at the moment but there is still much to improve. As we are working to upstream it we already fix some of the limitations that were not addressed so far. From what I see some of the things you mentioned could be added in the future. I think that the first set of changes should give more options for execution environment and memory allocation. Dynamic control over events would be next.
I imagine that the best way to go is to upstream it as close to the current state, collect list of required improvements, gradually improve the code. I won't be feasible to handle all at once.

> My point is that to ensure sane operation of this subsystem, you are practically forced to increase the heap size by the sum of the sizes of all the individual events

That is not true. You need to ensure there is enough memory to handle events that can occur simultaneously in the system. In most cases this will not be the sum of sizes of events of all types. In most cases this will not be related to number of event types.

> Regarding collecting sensor samples to a database which could be shared between the modules, i don't think it is the scope of event manager

I agree with @nordic-krch . I don't see this being related to event propagation. I would imagine a module that stores the samples (to RAM or even flash). It could ping listeners to query the DB for new values on change or even pass the pointer to the new value in the event. Anyway this should be module's code.

JordanYates commented:
> That is not true. You need to ensure there is enough memory to handle events that can occur simultaneously in the system. In most cases this will not be the sum of sizes of events of all types. In most cases this will not be related to number of event types.

It's not 100% true, but true enough (at least for how I would theoretically be using this library). 

> PR Description:
>> parts of the application functionality are separated into isolated modules that communicate with each other using events

By definition (isolated modules) most events are being generated asynchronously to each other. Assuming that because your N independent modules haven't all tried to generate events at the same time in the past they won't in the future seems like a bad assumption to me. You can't even work around it with some form of backoff mechanism because it hard reboots on failure (not that I think adding backoff mechanism would be a nice way to use this library).

Additionally, this design forces you to be ridiculously careful when changing any dynamic memory allocation anywhere in your program. Some library choosing to allocate 50 more bytes for whatever reason now means that any previous stability testing of the heap size is invalid.

I will accept that the last paragraph might just be my general aversion to `malloc` talking, but the earlier one is a serious problem IMO.

pdunaj commented:
> By definition (isolated modules) most events are being generated asynchronously to each other. Assuming that because your N independent modules haven't all tried to generate events at the same time in the past they won't in the future seems like a bad assumption to me.

That depends on the application. The one thing we can be sure that all assumptions depend on the application. It can happen that somebody would be willing to issue unbound number of events of the same or multiple types. If this is the case no system will handle it. It can happen that somebody will have a requirement to have large number of simultaneous events - and will have to provide memory for that. That is also true for any similar mechanism.

> Additionally, this design forces you to be ridiculously careful when changing any dynamic memory allocation anywhere in your program. Some library choosing to allocate 50 more bytes for whatever reason now means that any previous stability testing of the heap size is invalid.

It's true that using heap will not be ok for many people. Event if it is, one might not be willing to use the kernel heap. We know it and there is a plan to allow more generic approach to allocation.
I though that we will address that later - we cannot handle everything at the same time. There is however one cheap thing we could do.

@nordic-krch @zycz how about defining two weak function `event_manager_alloc` and `event_manager_free`? Being weak would allow user to overload them when needed. The cost of calls would be the same. Implementation cost is trivial so we could introduce it now. The default EM OOM behavior could be embedded in to `event_manager_alloc` instead of event allocators, allowing user to handle them as pleased.

zycz commented:
@pdunaj  defining two weak function event_manager_alloc and event_manager_free seems reasonable. Can add it in the PR  in which finall version of EM will be placed.

JordanYates commented:
> That depends on the application. The one thing we can be sure that all assumptions depend on the application.

We can agree on that :)

I'll wait to see what the final PR looks like before commenting any more on the design aspects.

My only other question is what is the larger picture here in terms of how this will be used.
Is the intent for other Zephyr subsystems to start using this to pass data around?
Is it simply an application level construct that other Zephyr code won't depend on?

Given our assumptions don't appear to line up, I don't really want to be forced to use this via its inclusion in other systems.

pdunaj commented:
> > That depends on the application. The one thing we can be sure that all assumptions depend on the application.
> 
> We can agree on that :)
> 
> I'll wait to see what the final PR looks like before commenting any more on the design aspects.
> 
> My only other question is what is the larger picture here in terms of how this will be used. Is the intent for other Zephyr subsystems to start using this to pass data around? Is it simply an application level construct that other Zephyr code won't depend on?
> 
> Given our assumptions don't appear to line up, I don't really want to be forced to use this via its inclusion in other systems.

I would leave it for application to decide as I believe this is something that belongs to application framework. Some like it event based, some prefer multiple threads or direct callbacks. Hard to say which framework will be best for a particular usecase, as it depends on application size, its complexity, system performance, memory size, etc.

If there will be event manager modules added to zephyr I would imagine these as additional layer overs specific functionality that can be used without EM.

nordic-krch commented on include/event_manager/event_manager_priv.h:
```diff
@@ -0,0 +1,256 @@
+		BUILD_ASSERT(offsetof(struct ename, header) == 0,	\
+				 "");					\
+		if (unlikely(!event)) {					\
+			printk("Event Manager OOM error\n");		\
```
i think this should be an option, typically when user uses dynamic allocation he need to take care about potential null return.
If you want to have failure here that should go through kernel api like `k_panic.

nashif replied:
agree, we should not be making decisions about reboots here. Also, using logging and not printk.

nordic-krch replied:
the reason why logging was not used here was because it is a macro so module that is using it may have no logging defined. This can be workarounded by calling an internal event_manager function which will use logging from known logging space.

nashif commented on include/event_manager/event_manager_priv.h:
```diff
@@ -0,0 +1,256 @@
+		BUILD_ASSERT(offsetof(struct ename, header) == 0,	\
+				 "");					\
+		if (unlikely(!event)) {					\
+			printk("Event Manager OOM error\n");		\
```
use logging

zycz commented:
Hi, I introduced changes we were talking about. Please rereview PR.

andyross commented on include/event_manager/event_manager_priv.h:
```diff
@@ -16,6 +16,11 @@
+#define _EM_FORCED_ALIGNMENT
+#else
+#define _EM_FORCED_ALIGNMENT __aligned(4)
+#endif
```
I've seen this pattern before (tripping over the difference between a C array and a linker-aligned array).  It's a known gotcha, and IIRC already dealt with by the STRUCT_SECTION_ITERABLE feature, which I'm guessing this code should be using?

What's the purpose to constructing your own linker section, basically?  Seems like we have a utility for that already.

pdunaj replied:
Hi @andyross , the x86 thing is caused by cache line optimizations for larger data. Maybe zephyr should give a konfig option to control it and disable it by default.

As for using the sections, the idea was to have event definitions decentralized. When module is compiled in it registers intention for event handling by adding it's object to the section specific for this event. Everything is done on build time so it also make us keep event callback arrays on flash. We use orphan sections for that.

I did not introduce linker script as I thought it would mean centralization of events (event manager would need to know about all the events in the system, and create right entries in the linker script).
However there is one thing I realized after you comment. At the moment Zephyr gives you the possibility to add fragments of linker script (zephyr_linker_sources). What I think we would need to do is to define a small linker script file for each event where right section placement is explicitly defined. There are other benefits of this as we could use sorting.

andyross reviewed (commented):
Drive-by nitpick, having noticed the weird-looking x86 commit title.

andyross commented:
@npitre may have more to say about ITERABLE

npitre commented:
> @npitre may have more to say about ITERABLE

Right. Use of `_EM_FORCED_ALIGNMENT` conditional on x86 looks suspicious.

I'd suggest investigating `Z_DECL_ALIGN()`, `STRUCT_SECTION_ITERABLE()` 
and `STRUCT_SECTION_FOREACH()`.

zycz commented:
Hi, I will make some changes in order to use Zephyr functions for dealing with sections.

Meanwhile:  @JordanYates , @nordic-krch  , @nashif could you please do rereview of things you requested to be changed?
Thanks

nordic-krch commented on include/event_manager/event_manager.h:
```diff
@@ -0,0 +1,317 @@
+
+#include <zephyr.h>
+#include <zephyr/types.h>
+#include <sys/reboot.h>
```
should not be needed

nordic-krch commented on include/event_manager/event_manager.h:
```diff
@@ -0,0 +1,317 @@
+#include <zephyr/types.h>
+#include <sys/reboot.h>
+#include <sys/__assert.h>
+#include <logging/log_ctrl.h>
```
should not be needed

nordic-krch commented on include/event_manager/event_manager.h:
```diff
@@ -0,0 +1,317 @@
+ *
+ * @brief Index of the highest subscriber priority level.
+ */
+#define SUBS_PRIO_MIN    _SUBS_PRIO_FIRST
```
since it is a public api it should be prefixed with EVENT_MANAGER, or at least obvious abbreviation (`EVT_MGR_`?)

nordic-krch replied:
if we use `app_event` approach that would become `APP_EVENT_SUBS_PRIO_MIN`

nordic-krch commented on include/event_manager/event_manager.h:
```diff
@@ -0,0 +1,317 @@
+#include <sys/__assert.h>
+#include <logging/log_ctrl.h>
+
+#include <event_manager/event_manager_priv.h>
```
i started to wonder if event_manager is not too generic name, especially that there are structures like `event_header`. Maybe it should be additionally prefixed with `app_`. `app_event` tells more than just `event`.

zycz replied:
Maybe lets first introduce event manager to Zephyr? We could thane change name.

nordic-krch commented on include/event_manager/event_manager.h:
```diff
@@ -0,0 +1,317 @@
+};
+
+
+/** @brief Event listener.
```
do we need `event_listener` `event_subscriber` and `event_type` to be in public header? Maybe they could be moved to priv header?

nordic-krch commented on include/event_manager/event_manager.h:
```diff
@@ -0,0 +1,317 @@
+};
+
+
+extern const struct event_listener __start_event_listeners[];
```
could those be moved to priv header?

nordic-krch commented on include/event_manager/event_manager.h:
```diff
@@ -0,0 +1,317 @@
+ *
+ * @param eh  Pointer to the event header element in the event object.
+ */
+void _event_submit(struct event_header *eh);
```
it probably could be moved to priv header.

nordic-krch commented on include/event_manager/event_manager.h:
```diff
@@ -0,0 +1,317 @@
+ * @param lname   Module name.
+ * @param cb_fn  Pointer to the event handler function.
+ */
+#define EVENT_LISTENER(lname, cb_fn) _EVENT_LISTENER(lname, cb_fn)
```
internal api's should be prefixed with `Z_`/`z_` instead of `_`

pdunaj replied:
@nordic-krch I am thinking how to approach it. So far we kept the changes in sync with NCS. This is needed as we don't have manpower to maintain this PR for long period of time while still maintaining NCS version. I don't think we can apply this change in NCS.
Maybe the renames can be done as a last step when we get to the point PR is accepted?

nordic-krch replied:
sure.

nordic-krch commented on include/event_manager/event_manager.h:
```diff
@@ -0,0 +1,317 @@
+#define EVENT_SUBMIT(event) _event_submit(&event->header)
+
+
+/** Initialize the Event Manager.
```
`@brief` missing

nordic-krch commented on include/event_manager/event_manager.h:
```diff
@@ -0,0 +1,317 @@
+ * The behavior of this function depends on the actual implementation.
+ * The default implementation of this function is no-operation.
+ * It is annotated as weak and is meant to be overridden by layer
+ * adding support for profiling mechanism.
```
lacking parameter description

nordic-krch commented on include/event_manager/event_manager.h:
```diff
@@ -0,0 +1,317 @@
+
+/** Initialize the Event Manager.
+ *
+ * @retval 0 If the operation was successful.
```
any error can be returned?

nordic-krch commented on include/event_manager/event_manager.h:
```diff
@@ -0,0 +1,317 @@
+ * The behavior of this function depends on the actual implementation.
+ * The default implementation of this function is no-operation.
+ * It is annotated as weak and is meant to be overridden by layer
+ * adding support for profiling mechanism.
```
`@param` missing

nordic-krch commented on include/event_manager/event_manager.h:
```diff
@@ -0,0 +1,317 @@
+ * The behavior of this function depends on the actual implementation.
+ * The default implementation of this function is no-operation.
+ * It is annotated as weak and is meant to be overridden by layer
+ * adding support for profiling mechanism.
```
`@retval`

nordic-krch commented on include/event_manager/event_manager.h:
```diff
@@ -0,0 +1,317 @@
+ * @param size  Amount of memory requested (in bytes).
+ * @retval Address of the allocated memory if successful, otherwise NULL.
+ **/
+void *event_manager_alloc(size_t size);
```
is this function public or used only in the wrapper macro?

zycz replied:
It is only here because it can be overwritten by the user.

nordic-krch commented on include/event_manager/event_manager.h:
```diff
@@ -0,0 +1,317 @@
+void *event_manager_alloc(size_t size);
+
+
+/** Free memory occupied by the event.
```
`@brief`

nordic-krch reviewed (commented):
few comments to `event_manager.h`. Will continue tomorrow.

nordic-krch commented on include/event_manager/event_manager_priv.h:
```diff
@@ -0,0 +1,228 @@
+			(struct ename *)event_manager_alloc(sizeof(*event));	\
+		BUILD_ASSERT(offsetof(struct ename, header) == 0,		\
+				 "");						\
+		event->header.type_id = _EVENT_ID(ename);			\
```
`if (event != NULL)` is missing

nordic-krch commented on include/event_manager/event_manager_priv.h:
```diff
@@ -0,0 +1,228 @@
+				 sizeof(*event), "");					\
+		BUILD_ASSERT(offsetof(struct ename, header) == 0,			\
+				 "");							\
+		event->header.type_id = _EVENT_ID(ename);				\
```
missing check if event was successfully allocated

nordic-krch commented on include/event_manager/event_manager.h:
```diff
@@ -0,0 +1,317 @@
+	/** Pointer to the function that is called when an event
+	 *  is handled.
+	 */
+	bool (*notification)(const struct event_header *eh);
```
it's missing explanation what `bool` return value means. It would be good to extract this into typedef with proper doxygen.

nordic-krch commented on subsys/event_manager/event_manager.c:
```diff
@@ -0,0 +1,233 @@
+	void *event = k_malloc(size);
+
+	if (unlikely(!event)) {
+		printk("Event Manager OOM error\n");
```
use log here

nordic-krch commented on subsys/event_manager/event_manager.c:
```diff
@@ -0,0 +1,233 @@
+
+	if (unlikely(!event)) {
+		printk("Event Manager OOM error\n");
+		LOG_PANIC();
```
it's redundant, done in k_panic

nordic-krch commented on subsys/event_manager/event_manager.c:
```diff
@@ -0,0 +1,233 @@
+	}
+
+	if (et->log_event) {
+		char log_buf[CONFIG_EVENT_MANAGER_EVENT_LOG_BUF_LEN];
```
This way of logging is not very optimal (sprintf + logging strings are much slower than logging values). Wonder if it could be reworked so that event logging function is logging directly.
I will come up with a proposal just need to check it.

nordic-krch replied:
here is my proposal:
https://github.com/nordic-krch/zephyr/commits/event_manager_logging
you can find two commits there, one with logging proposal. It enables logging directly using logging API and not doing sprintf. In second commit i tried to compare CPU load impact. Calculated load for things other than idle and logging thread to get load of application. I got `0.45%` with new logging and `0.66%` without. So seems like app was spending 33% of time sprintf-logging.

zycz replied:
Thanks for your input. I thought that it should be backward compatible with the previous version.  Do you thing Kconfig option would be ok in this case?

nordic-krch replied:
it could be but since current version does not have any benefits over this proposal it should be eventually deprecated and removed. Prototype of logging function shall be changed when newer version is used.

pdunaj replied:
Please don't reuse the old name `log_event`. We should create a new member of the structure.

This is simple change (our shame we have not done macros hiding logging in the first place when we migrated the event manager out of nrf_desktop), bigger ones are introduced by zephyr every quarter.
Nevertheless adding backward compatibility should be very simple. You can add a config option and handle log either through old `log_event` or new `log_fn` structure member. Config option could make only one of them available (preferred option since we would like to drop the old ways).

pdunaj replied:
Have another thought about it. We cannot do it in any other way then proposed by @nordic-krch . Adding a new member would required to change the macro defining an event (so we can assign right member with a given logging callback). Doing that conditionally based on kconfig option is not really safe as at some point there will be a conflict between function given and expected.
@nordic-krch proposal using the old api gives an option for explicit check on the compatibility by passing the null pointer as buffer. So let's do it as was proposed.

We could still keep the simple backward compatibility at event_manager.c - if enabled buffer is created and passed to the function as it used to be. New functions should handle it fine. @zycz , let's add this minimum backward compatibility option and inform in change log that old ways will be dropped.

nordic-krch reviewed (commented):
few comments

zycz committed:
event_manager: Event manager cleanup

Add changes cleaning event manager.

Signed-off-by: Jan Zyczkowski <jan.zyczkowski@nordicsemi.no>

zycz committed:
event_manager: Add subscribe first to event

Add first subscriber to event functionality to event manager.

Jira: NCSDK-9067

Signed-off-by: Jan Zyczkowski <jan.zyczkowski@nordicsemi.no>

zycz committed:
event manager: tests: tests for event manager

Add unit tests for event manager.

Signed-off-by: Jan Zyczkowski <jan.zyczkowski@nordicsemi.no>

pdunaj commented:
Seems like all changes are done. I think the last outstanding thing is the matter of name. I don't personally mind the current one, especially that many people are familiar with it :)

Hi @nordic-krch do you think we can wrap up the PR and have it merged?
Anybody?

nordic-krch commented on include/event_manager/event_manager_priv.h:
```diff
@@ -0,0 +1,304 @@
+	bool init_log_enable;
+
+	/** Function to log data from this event. */
+	int (*log_event)(const struct event_header *eh, char *buf,
```
Lets assume that we would like to use only new logging method using `EVENT_MAANGER_LOG` macro in zephyr. Legacy method is only for backward compatibility in NCS. I would propose following approach: 
- Deprecate legacy logging method (the one less efficient using snprintf)
- Create here two fields with two typedefs. New function will be `void (*log_event_func_t)(const struct event_header *eh)`
- Use `_Generic` keyword do detect type of function provided by the user in the macro and assign to one field and leave another as null
- in processing you pick method based on which pointer is not null.

If `_Generic` keyword is not supported by the compiler legacy method cannot be used and that should be fine because it will only be used in NCS and all cortex_m compilers that we are aware of support `_Generic`.

Detection of `_Generic` support: https://github.com/zephyrproject-rtos/zephyr/blob/731241f8d049924591cf4f46dda26ddd0782d8c7/include/sys/cbprintf.h#L28

Examples of `_Generic` use: https://github.com/zephyrproject-rtos/zephyr/blob/731241f8d049924591cf4f46dda26ddd0782d8c7/include/sys/cbprintf_internal.h#L91

nordic-krch commented on include/event_manager/event_manager.h:
```diff
@@ -0,0 +1,248 @@
+ * @param ... `printf`- like format string and variadic list of arguments corresponding to
+ *            the format string.
+ */
+#define EVENT_MANAGER_LOG(eh, ...) do {								\
```
Would be good to add usage in the sample.

nordic-krch commented on samples/subsys/event_manager/src/events/measurement_event.c:
```diff
@@ -0,0 +1,24 @@
+{
+	struct measurement_event *event = cast_measurement_event(eh);
+
+	return snprintf(buf, buf_len, "val1=%d val2=%d val3=%d", event->value1,
```
Use new logging method

nordic-krch commented on samples/subsys/event_manager/src/events/config_event.c:
```diff
@@ -0,0 +1,23 @@
+{
+	struct config_event *event = cast_config_event(eh);
+
+	return snprintf(buf, buf_len, "init_val_1=%d", event->init_value1);
```
use new method

nordic-krch commented on tests/subsys/event_manager/testcase.yaml:
```diff
@@ -0,0 +1,4 @@
+tests:
+  event_manager.core:
+    arch_exclude: xtensa
```
what is the reason for not supporting xtensa? Is there a plan to fix it?

zycz replied:
There is a problem with xtensa linker script. It lacks of: #include <snippets-sections.ld>in  soc/xtensa/sample_controller/linker.ld
And because of it compiler does not see linker script added in cmake.

zycz replied:
Do you know anyone who could comment how we should procced with this problem?

nordic-krch replied:
Open an issue, they will assign xtensa maintainer

nordic-krch commented on samples/subsys/event_manager/README.rst:
```diff
@@ -0,0 +1,85 @@
+  The module calculates and logs basic statistics about one of the measurement values that are transmitted as part of the event.
+
+
+Requirements
```
you can remove that section as it is now supported in general (except for xtensa)

nordic-krch commented on samples/subsys/event_manager/README.rst:
```diff
@@ -0,0 +1,85 @@
+Testing
+=======
+
+After programming the sample to your development kit, test it by performing the following steps:
```
Maybe you could rephrase it a bit so it suites qemu platforms

nordic-krch commented on samples/subsys/event_manager/README.rst:
```diff
@@ -0,0 +1,85 @@
+
+Building and running
+********************
+Build and flash Event Manager as follows, changing nrf52840dk_nrf52840 for your board:
```
I think that you can remove that section as well.

nordic-krch commented on samples/subsys/event_manager/prj.conf:
```diff
@@ -0,0 +1,17 @@
+# Logger configuration
+CONFIG_LOG=y
+CONFIG_LOG_DEFAULT_LEVEL=2
+CONFIG_LOG_BACKEND_UART=y
```
can be removed, on most platforms enabling logging comes with default backend and for some it's not uart.

nordic-krch commented on samples/subsys/event_manager/sample.yaml:
```diff
@@ -0,0 +1,13 @@
+    build_only: true
+    arch_exclude: xtensa
+    integration_platforms:
+      - nrf52dk_nrf52832
```
use something more generic here,e.g. `qemu_x86`. It is typically used in zephyr (or `native_posix`)

jfischer-no reviewed (commented):
I have looked at this PR more closely, among other because I wanted something similar for USB subsystem. Event Manager as it is might be robust and fast, but that is about it. It is probably not extensible and not extensible at all to runtime subscription/unsubscription. Macro acrobatics hurt; need of EVENT_TYPE_DEFINE and EVENT_TYPE_DECLARE is awkward, especially since event define header must also be included by listener, and the latter could pull the information even without macro obfuscation; no chance to get rid of LINKER_ORPHAN_SECTION. I have tried to improve and change it, but went over to rewrite something similar from scratch. I will come soon with a proposal.

pdunaj commented:
> I have looked at this PR more closely, among other because I wanted something similar for USB subsystem. Event Manager as it is might be robust and fast, but that is about it. It is probably not extensible and not extensible at all to runtime subscription/unsubscription. Macro acrobatics hurt; need of EVENT_TYPE_DEFINE and EVENT_TYPE_DECLARE is awkward, especially since event define header must also be included by listener, and the latter could pull the information even without macro obfuscation; no chance to get rid of LINKER_ORPHAN_SECTION. I have tried to improve and change it, but went over to rewrite something similar from scratch. I will come soon with a proposal.

> It is probably not extensible and not extensible at all to runtime subscription/unsubscription.

We can add enable/disable if there is a need for that. Runtime subscription costs I am not sure it is really needed in embedded we work on.

> Macro acrobatics hurt; need of EVENT_TYPE_DEFINE and EVENT_TYPE_DECLARE is awkward

Depends what you are looking for. Performance was priority. Defining new event is not really magic. We have never got any question about it and we know people use it.

>  no chance to get rid of LINKER_ORPHAN_SECTION

Please check again. It was removed without a problem.

jfischer-no commented on subsys/event_manager/event_manager.c:
```diff
@@ -0,0 +1,237 @@
+	sys_slist_t events = SYS_SLIST_STATIC_INIT(&events);
+
+	/* Make current event list local. */
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	if (sys_slist_is_empty(&eventq)) {
+		k_spin_unlock(&lock, key);
+		return;
+	}
+
+	sys_slist_merge_slist(&events, &eventq);
+
+	k_spin_unlock(&lock, key);
```
The same can be implemented in one-line using `struct k_fifo` and `k_fifo_get()`. I also see no advantages not to do it this way, very unlikely `eventq` would contain more than a single event per `k_work_submit(&event_processor);`.

pdunaj replied:
Maybe. It could be a potential change for later.
Originally it was implemented to be as fast as possible so heavier api was avoided. To change it we would need to see if there would be no impact on performance.

zycz committed:
event manager: renamed event manager files name

Renamed event_manager files to app_event_manager.

Signed-off-by: Jan Zyczkowski <jan.zyczkowski@nordicsemi.no>

nordic-krch commented on include/app_event_manager/app_event_manager.h:
```diff
@@ -0,0 +1,422 @@
+ *
+ * @param id  ID.
+ */
+#define ASSERT_APP_EVENT_ID(id) \
```
APP_EVENT_ASSERT_ID

nordic-krch commented on include/app_event_manager/app_event_manager.h:
```diff
@@ -0,0 +1,422 @@
+ */
+#define APP_EVENT_HOOK_POSTPROCESS_REGISTER(hook_fn)	\
+	_APP_EVENT_HOOK_POSTPROCESS_REGISTER(hook_fn,	\
+	_EM_SUBS_PRIO_ID(_EM_SUBS_PRIO_NORMAL))
```
_EM should probably be changed to Z_APP_EVENT

zycz committed:
event_manager: Event manager cleanup v2

Add changes cleaning event manager.

Signed-off-by: Jan Zyczkowski <jan.zyczkowski@nordicsemi.no>

nordic-krch commented on include/app_event_manager/app_event_manager.h:
```diff
@@ -0,0 +1,422 @@
+ * @param hook_fn Hook function.
+ */
+#define APP_EVENT_HOOK_POSTPROCESS_REGISTER_LAST(hook_fn)                      \
+	const struct {} __event_hook_postprocess_last_sub_redefined = {};  \
```
what is the purpose of this variable?

zycz replied:
It is a marker indicating that only one hook can be registered as Last.

nordic-krch reviewed (commented):
Can you add test to hooks  `APP_EVENT_HOOK...`. A lot of those is defined in the API but they are nowhere used.

pdunaj commented on include/event_manager/event_manager.h:
```diff
@@ -0,0 +1,238 @@
+ * @param ename  Name of the event.
+ */
+#define EVENT_SUBSCRIBE_FIRST(lname, ename)							\
+	_EVENT_SUBSCRIBE(lname, ename, _EM_MARKER_FIRST_ELEMENT);				\
```
There is perfect sense. These items should not be used directly - placing them in priv header makes it explicit.

pdunaj commented on include/event_manager/event_manager.h:
```diff
@@ -0,0 +1,238 @@
+ *
+ * @param event  Pointer to the event object.
+ */
+#define EVENT_SUBMIT(event) _event_submit(&event->header)
```
simplification of usage - I recommend you to look at the code again

pdunaj commented on include/event_manager/event_manager.h:
```diff
@@ -0,0 +1,238 @@
+ * @param ev_info_struct   Data structure describing the event type.
+ *                         NULL can be used in case no data structure is needed.
+ */
+#define EVENT_TYPE_DEFINE(ename, init_log_en, log_fn, ev_info_struct) \
```
Sure you don't ;)

pdunaj commented on include/event_manager/event_manager_priv.h:
```diff
@@ -0,0 +1,324 @@
+	log_event_data_dep log_event_func_dep;
+
+	/** Custom data related to tracking. */
+	const void *trace_data;
```
We use it in NCS

pdunaj commented on subsys/event_manager/Kconfig:
```diff
@@ -0,0 +1,56 @@
+	help
+	  Enable Event Manager.
+	  Note that Event Manager uses orphan sections to handle its
+	  data structures.
```
Not any more. This is a leftover after cleanup (@zycz )

pdunaj commented on subsys/event_manager/event_manager.c:
```diff
@@ -0,0 +1,237 @@
+	BUILD_ASSERT(CONFIG_EVENT_MANAGER_EVENT_LOG_BUF_LEN != 1);
+	char log_buf[CONFIG_EVENT_MANAGER_EVENT_LOG_BUF_LEN];
+
+	int pos = et->log_event_func_dep(eh, log_buf, sizeof(log_buf));
```
I accept you have a different view on this but I don't agree. There is no rule for that and there are many well known advantages of defining variables as close to code as possible (i.e. in place).

pdunaj commented on include/event_manager/event_manager.h:
```diff
@@ -0,0 +1,238 @@
+ *
+ * @param ename  Name of the event.
+ */
+#define EVENT_TYPE_DECLARE(ename) _EVENT_TYPE_DECLARE(ename)
```
That seemed the simplest to do at the time when event manager was created. Can you elaborate on proposal?

Note that we migrate event manager from NCS - it's not a new code. There is quite a legacy behind. If there is no real advantage I would not change anything. The current API works fine.

zycz commented:
App Event Manager right now won't be up streamed to Zephyr. I close PR.

pdunaj commented:
hi @rodrigobrochado , I have noticed your emoji. Any reason? The code is still available in Nordic Connect SDK :)

hongshui3000 commented:
It's a pity that this PR was closed

gmarull commented on doc/reference/kernel/other/event_manager.rst:
```diff
@@ -0,0 +1,359 @@
+To initialize the Event Manager, complete the following steps:
+
+1. Include :file:`event_manager.h` in your :file:`main.c` file.
+#. Call :c:func:`event_manager_init()`.
```
why does event manager need to be started manually? could it be done via SYS_INIT?

pdunaj replied:
The event manager was created to be an application skeleton (not a system one) and I believe it is better for application to have a control when event processing is started.
This said initialization function does almost nothing at the moment - it performs checks but was left in case we would like to extend the configuration in the future (like start a separate event processing thread, prepare an allocator or something similar).

gmarull commented on doc/reference/kernel/other/event_manager.rst:
```diff
@@ -0,0 +1,359 @@
+.. code-block:: c
+
+	/* Allocate event. */
+	struct sample_event *event = new_sample_event();
```
is this dynamic allocation?

pdunaj replied:
Yes - there are hooks (weak linkage) that allows user to pass own allocator functions. by default it's k_malloc.

gmarull commented on include/event_manager/event_manager_priv.h:
```diff
@@ -0,0 +1,220 @@
+	const struct event_subscriber _CONCAT(_CONCAT(__event_subscriber_, ename), lname) __used	\
+	__attribute__((__section__(_EVENT_SUBSCRIBERS_SECTION_NAME(ename, prio)))) = {			\
+		.listener = &_CONCAT(__event_listener_, lname),						\
+	}
```
any reason not to use https://docs.zephyrproject.org/latest/kernel/iterable_sections/index.html ?

pdunaj replied:
I think I had a look during the last rework if we can use more of standard macros. There must have been something that prevented it but I don't recall what it was.

gmarull commented on include/event_manager/event_manager_priv.h:
```diff
@@ -0,0 +1,220 @@
+ * ename type.
+ */
+#define _EVENT_ALLOCATOR_FN(ename)						\
+	static inline struct ename *_CONCAT(new_, ename)(void)			\
```
I think all event manager function calls should be namespaced, this just creates `new_...`.

pdunaj replied:
you are right - this is leftover from times it was used in one app only that was not fixed during API rework, we should probably follow up on that in NCS.

pdunaj replied:
@zycz , can you create a ticket to follow this up?

gmarull commented on include/event_manager/event_manager_priv.h:
```diff
@@ -0,0 +1,220 @@
+ * ename type.
+ */
+#define _EVENT_ALLOCATOR_FN(ename)						\
+	static inline struct ename *_CONCAT(new_, ename)(void)			\
```
is dynamic allocation a mandatory thing?

pdunaj replied:
How else you would want to do it?
Events are dynamic so you have to do dynamic allocation on them. I expect having a preset of event objects create at build time would be quite limiting.
As for an actual allocator that is used - by default it is heap (k_malloc heap to be precise) but user can hook any other allocator (slab for example).

gmarull commented on include/event_manager/event_manager_priv.h:
```diff
@@ -0,0 +1,220 @@
+						(ename, _SUBS_PRIO_ID(_SUBS_PRIO_FINAL)),	\
+		},										\
+		.init_log_enable		= init_log_en,					\
+		.log_event			= (IS_ENABLED(CONFIG_LOG) ? (log_fn) : (NULL)),	\
```
pointer should likely be excluded if CONFIG_LOG=n, so we save some space

gmarull commented on include/event_manager/event_manager_priv.h:
```diff
@@ -0,0 +1,220 @@
+			[_SUBS_PRIO_FIRST]	= _EVENT_SUBSCRIBERS_START			\
+						(ename, _SUBS_PRIO_ID(_SUBS_PRIO_FIRST)),	\
+			[_SUBS_PRIO_NORMAL]	= _EVENT_SUBSCRIBERS_START			\
+						(ename, _SUBS_PRIO_ID(_SUBS_PRIO_NORMAL)),	\
```
general comment: would be good to run clang-format, I see multiple inconsistencies all around

pdunaj replied:
Looking at this snipped I think this is an old implementation before the rework. Please check the one in NCS for the newest code.
It might be that many things seen here are already gone.

gmarull commented on subsys/event_manager/event_manager.c:
```diff
@@ -0,0 +1,233 @@
+{
+}
+
+void * __weak event_manager_alloc(size_t size)
```
I think that in general we should avoid using weaks (it's a non-standard feature). Kconfig options to allow custom re-implementations would be a better choice IMO.

gmarull commented on subsys/event_manager/event_manager_shell.c:
```diff
@@ -0,0 +1,173 @@
+	for (const struct event_type *et = __start_event_types;
+	     (et != NULL) && (et != __stop_event_types); et++) {
+
+		size_t ev_id = et - __start_event_types;
```
we should likely avoid using such variables directly and provide iterator functions instead, see for example `device_required_foreach`.

gmarull commented on samples/subsys/event_manager/src/events/ack_event.h:
```diff
@@ -0,0 +1,36 @@
+ * @{
+ */
+
+#include "event_manager/event_manager.h"
```
let's use headers correctly: anything in the include path (non-local) must use `<>`.

gmarull commented on samples/subsys/event_manager/src/events/ack_event.h:
```diff
@@ -0,0 +1,36 @@
+
+#ifdef __cplusplus
+}
+#endif
```
does this need to run on C++ code?

pdunaj replied:
I expect sample simply copied code from some previously defined event. Is there any harm coming out of it?

gmarull commented on samples/subsys/event_manager/src/events/config_event.c:
```diff
@@ -0,0 +1,22 @@
+
+static void log_config_event(const struct event_header *eh)
+{
+	struct config_event *event = cast_config_event(eh);
```
maybe a macro would be more useful/readable (aligning with `CONTAINER_OF` usage), e.g.

```c
struct config_event *event = EVM_CAST(eh, config);
```

pdunaj replied:
good idea - we could follow this up along with new_event api @zycz

gmarull reviewed (commented):
some more comments. I have to say that except a few rough edges I like what is presented in this PR.
